pyfr/plugins/tavg.py

```diff
@@ -129,23 +156,69 @@ def _eval_acc_exprs(self, intg):
                         subs[f'grad_{pname}_{dim}'] = grad
 
             # Evaluate the expressions
-            exprs.append([npeval(v, subs) for v in self.aexprs])
+            exprs.append(np.array([npeval(v, subs) for v in self.aexprs]))
 
-        # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
+        return exprs
 
-    def _eval_fun_exprs(self, intg, accex):
-        exprs = []
+    def _eval_fun_exprs(self, avars):
 
-        # Iterate over each element type our averaging region
-        for avals in accex:
-            # Prepare the substitution dictionary
-            subs = dict(zip(self.anames, avals.swapaxes(0, 1)))
+        # Prepare the substitution dictionary
+        subs = dict(zip(self.anames, avars))
+    
+        # Evaluate the function and return
+        return np.array([npeval(v, subs) for v in self.fexprs])
 
-            exprs.append([npeval(v, subs) for v in self.fexprs])
+    def _eval_fun_var(self, dev, accex):
+        dfexpr, exprs = [], []
+        dh, an = np.sqrt(self.delta_h), self.anames
 
-        # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
+        # Iterate over each element type our averaging region
+        for av in accex:
+            df = []
+
+            # Evaluate the function
+            fx = self._eval_fun_exprs(av)
+            exprs.append(fx)
+
+            for i in range(len(an)):
+                # Calculate step size
+                h = dh * np.maximum(abs(av[i]), dh, where=abs(av[i])>dh, out=np.ones_like(av[i]))                
+                av[i] += h
+                
+                # Evaluate function for the step
+                fxh = self._eval_fun_exprs(av)
+                
+                # Calculate derivatives for functional averages
+                df.append((fxh - fx) / h)
+                av[i] -= h
+
+            # Stack derivatives     
+            dfexpr.append(np.array(df))
+
+        # Multiply by variance and take RMS value
+        fv = [np.linalg.norm(df * sd[:, None], axis=0) for df, sd in zip(dfexpr, dev)]
+        
+        return exprs, fv 
+
+    def _acc_avg_var(self, intg, currex):  
+        prevex, vaccex, accex = self.prevex, self.vaccex, self.accex  
+
+        # Weights for online variance and average
+        Wmp1mpn = intg.tcurr - self.prevt
+        W1mpn = intg.tcurr - self.tstart_acc
+        Wp = 2 * (W1mpn - Wmp1mpn) * W1mpn
+
+        # Iterate over element type
+    
+        for v, a, p, c in zip(vaccex, accex, prevex, currex):
+            ppc = p + c
+            # Accumulate average
+            a += Wmp1mpn * ppc
+
+            # Accumulate variance          
+            v += Wmp1mpn*(c**2 + p**2 - 0.5 * ppc**2) 
+            if self.tstart_acc != self.prevt:
+                v +=  (Wmp1mpn / Wp * (a - W1mpn * ppc)**2)
```

Comment:
No point recomputing `Wmp1mpn / Wp` each iteration, redefine `Wmp1mpn` to include a factor of `1/Wp`.